### PR TITLE
[SG2] Implement add/remove blackboard items

### PR DESCRIPTION
**Pull request by dhsavell**
### Purpose of this PR

This PR enables the Blackboard add menu and adds support for GraphType-backed variables.

Note: color and matrix fields display as vectors pending further exploration of field editors.

image

image

---
### Testing status

- Blackboard view exists when graph is opened
- Variable types are presented when pressing the Properties "+" button - C1007043
- When adding a variable, the "Initialization" field matches its type (i.e. a vec4 shows a vec4 field, bool shows a checkbox) - C1007049
  - Caveat: color, matrix 2/3/4, see above
- Blackboard items can be selected, rearranged, and deleted - C1007046, C1007054
  - Caveat: underlying data connection is not deleted

---
### Comments to reviewers

There are a few interesting things I called out in comments below.

I was working in `ShaderGraphExampleTypes` so I took a moment to clean out some of the dead code there. In particular, we were maintaining a mapping of names to type handles that was unused. However, if these changes could cause issues, please let me know and I'll revert them.

### CODE REVIEW

File: com.unity.shadergraph/Editor/GraphUI/DataModel/Examples/ShaderGraphExampleTypes.cs [hunk 1]
```diff
@@ -64,44 +53,39 @@ public static TypeHandle GetGraphType(PortHandler reader)
...
-            { "Color", Color },
-            { "AnimationClip", AnimationClip },
-            { "Mesh", Mesh },
-            { "Texture2D", Texture2D },
-            { "Texture3D", Texture3D },
-            { "DayOfWeek", DayOfWeek },
-            { "Gradient", GradientTypeHandle }
-        };
-
-        public static IEnumerable<string> TypeHandleNames => TypeHandlesByName.Keys;
```
Comment (dhsavell) [inline com.unity.shadergraph/Editor/GraphUI/DataModel/Examples/ShaderGraphExampleTypes.cs:104]: None of this is actually used - it was used for old exploration stuff (`AddPortPart`, which is now entirely commented out)

File: com.unity.shadergraph/Editor/GraphUI/DataModel/Examples/ShaderGraphExampleTypes.cs [hunk 2]
```diff
@@ -218,7 +202,7 @@ public object ObjectValue
                     default:
                         switch (GetPrimitive())
                         {
-                            case GraphType.Primitive.Int: sc(0, (float)value); return;
+                            case GraphType.Primitive.Int: sc(0, Convert.ToSingle(value)); return;
```
Comment (dhsavell) [inline com.unity.shadergraph/Editor/GraphUI/DataModel/Examples/ShaderGraphExampleTypes.cs:205]: Fun type headache: if we get an `int` as an `object`, we can't directly cast it to `float`.

**Comment by dhsavell** [comment]
Weird... I had that happen once but thought it was a one-off because my graph was in a broken state after a reload. If it's surfacing again, there's probably something more to it. I'll update the associated TestRail case with a note.

File: com.unity.shadergraph/Editor/GraphUI/DataModel/GraphDataVariableNodeModel.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+namespace UnityEditor.ShaderGraph.GraphUI
+{
+    public class GraphDataVariableNodeModel : VariableNodeModel, IGraphDataPortModelOwner
+    {
+        [SerializeField]
+        string m_GraphDataName;
+
+        /// <summary>
+        /// Graph data name associated with this node. If null, this node is a searcher preview with type determined
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/GraphUI/DataModel/GraphDataVariableNodeModel.cs:17]: ah crap need to get rid of this bit with searcher preview

File: com.unity.shadergraph/Editor/GraphUI/DataModel/GraphDataVariableNodeModel.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        /// by the registryKey property.
+        /// </summary>
+        public string graphDataName
+        {
+            get => m_GraphDataName;
+            set => m_GraphDataName = value;
+        }
+
+        /// <summary>
+        /// This node's registry key. If graphDataName is set, this is read from the graph. Otherwise, it is set
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/GraphUI/DataModel/GraphDataVariableNodeModel.cs:27]: same with this bit of the comment regarding setPreviewRegistryKey

File: com.unity.shadergraph/Editor/GraphUI/DataModel/IGraphDataPortModelOwner.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿using UnityEditor.GraphToolsFoundation.Overdrive.BasicModel;
+using UnityEditor.ShaderGraph.GraphDelta;
+
+namespace UnityEditor.ShaderGraph.GraphUI
+{
+    public interface IGraphDataPortModelOwner
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/GraphUI/DataModel/IGraphDataPortModelOwner.cs:6]: Will document this as its a public class

File: com.unity.shadergraph/Editor/GraphUI/DataModel/ShaderGraphModel.cs [hunk 6]
```diff
@@ -276,5 +300,26 @@ public static bool ShouldElementBeVisibleToSearcher(ShaderGraphModel shaderGraph
...
+
+        public override IVariableNodeModel CreateVariableNode(IVariableDeclarationModel declarationModel,
+            Vector2 position, SerializableGUID guid = default, SpawnFlags spawnFlags = SpawnFlags.Default)
+        {
+            Action<VariableNodeModel> initCallback = variableNodeModel =>
+            {
+                if (declarationModel is GraphDataVariableDeclarationModel model && variableNodeModel is GraphDataVariableNodeModel graphDataVariable)
+                {
+                    variableNodeModel.VariableDeclarationModel = model;
+                    // Every time a variable node is added to the graph, add a node referencing that variable on the CLDS level
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/GraphUI/DataModel/ShaderGraphModel.cs:312]: should amend this comment to: "add a reference node pointing back to the variable/property that is wrapped by the VariableDeclarationModel, on the CLDS level"

File: com.unity.shadergraph/Editor/GraphUI/GraphElements/ModelUI/Blackboard/ShaderGraphSectionModel.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using UnityEditor.GraphToolsFoundation.Overdrive.BasicModel;
+
+namespace UnityEditor.ShaderGraph.GraphUI
+{
+    public class ShaderGraphSectionModel : SectionModel
```
Comment (Nightmask3) [inline com.unity.shadergraph/Editor/GraphUI/GraphElements/ModelUI/Blackboard/ShaderGraphSectionModel.cs:5]: In theory we could remove this rn cause it doesnt do anything but we might need it later, so leaving it here for now

**Review by brett-groshong — APPROVED** [review]

**Pull request merged**